assets/js/admin/marketplace-suggestions.js

```diff
@@ -0,0 +1,340 @@
+( function( $, marketplace_suggestions ) {
+	$( function() {
+		if ( 'undefined' === typeof marketplace_suggestions ) {
+			return;
+		}
+
+		// Stand-in wcSettings.recordEvent for until tracks PR is merged.
+		// https://github.com/woocommerce/woocommerce/pull/22746
+		window.wcSettings = window.wcSettings || {};
+		window.wcSettings.recordEvent = window.wcSettings.recordEvent  || function() { };
+
+		// Tracks events sent in this file:
+		// - marketplace_suggestion_displayed
+		// - marketplace_suggestion_clicked
+		// - marketplace_suggestion_dismissed
+		// All are prefixed by {WC_Tracks::PREFIX}.
+		// All have one property for `suggestionSlug`, to identify the specific suggestion message.
+
+		function dismissSuggestion( suggestionSlug ) {
+			// hide the suggestion in the UI
+			var selector = '[data-suggestion-slug=' + suggestionSlug + ']';
+			$( selector ).fadeOut();
+
+			// save dismissal in user settings
+			jQuery.post(
+				ajaxurl,
+				{
+					'action': 'add_dismissed_marketplace_suggestion',
+					'_wpnonce': marketplace_suggestions.dismiss_suggestion_nonce,
+					'slug': suggestionSlug,
+				}
+			);
+
+			window.wcSettings.recordEvent( 'marketplace_suggestion_dismissed', {
+				suggestionSlug: suggestionSlug
+			} );
+		}
+
+		function renderDismissButton( suggestionSlug ) {
+			var dismissButton = document.createElement( 'a' );
+
+			dismissButton.classList.add( 'suggestion-dismiss' );
+			dismissButton.setAttribute( 'href', '#' );
+			dismissButton.onclick = function( event ) {
+				event.preventDefault();
+				dismissSuggestion( suggestionSlug );
+			}
+
+			return dismissButton;
+		}
+
+		function renderLinkout( slug, url, text, isButton ) {
+			var linkoutButton = document.createElement( 'a' );
+
+			linkoutButton.setAttribute( 'href', url );
+			linkoutButton.setAttribute( 'target', 'blank' );
+			linkoutButton.textContent = text;
+
+			linkoutButton.onclick = function( event ) {
+				window.wcSettings.recordEvent( 'marketplace_suggestion_clicked', {
+					suggestionSlug: slug
+				} );
+			}
+
+			if (isButton) {
+				linkoutButton.classList.add( 'button' );
+			}
+			else {
+				linkoutButton.classList.add( 'linkout' );
+				var linkoutIcon = document.createElement( 'span' );
+				linkoutIcon.classList.add( 'dashicons', 'dashicons-external' );
+				linkoutButton.appendChild( linkoutIcon );
+			}
+
+			return linkoutButton;
+		}
+
+		function renderSuggestionIcon( slug, iconUrl ) {
+			if ( !iconUrl ) {
+				return null;
+			}
+
+			var image = document.createElement( 'img' );
+			image.src = iconUrl;
+			image.classList.add( 'marketplace-suggestion-icon' );
+
+			return image;
+		}
+
+		function renderSuggestionContent( slug, title, copy ) {
+			var left = document.createElement( 'div' );
+
+			left.classList.add( 'marketplace-suggestion-container-content' );
+
+			if ( title ) {
+				var titleHeading = document.createElement( 'h4' );
+				titleHeading.textContent = title;
+				left.appendChild( titleHeading );
```

Comment:
Perhaps some escaping of the data from the JSON should be done here to make sure only text is output and not HTML. There are some workarounds for that like appending the raw content to a div and using `.text()` to grab the safe non-html text.

Reply:
We're protected from this because we're assigning `textContent` – any bad stuff gets escaped for us:

<img width="858" alt="screen shot 2019-02-21 at 3 33 40 pm" src="https://user-images.githubusercontent.com/4167300/53139311-1c03a400-35ee-11e9-9734-8fa93f357e08.png">